cca_in_ad -> replace kubernetes/docker launchers with container launcher

djmitche opened the pull request:
The kubernetes and docker launchers have overlapping functionality, and
both have serious bugs and missing functionality.  The container
launcher implements the same behavior, with fewer bugs and more
functionality.  In particular, it

 - handles config from files (conf.d) for containers under kubernetes
 - handles changes to pod annotations
 - correctly logs from file or socket based on `logs_config.{k8s,docker}_container_use_file`
 - captures logs under containerd when `k8s_container_use_file` is false

The entire container launcher is built behind the 'docker' build tag,
which really means "containers" and not specifically docker.

Outside of the situations where they have obvious bugs, this launcher
attempts to preserve behavior of the kubernetes and docker launchers in
the details, such as defaults for the "Source" and "Service" properties.
It does not preserve those launchers' behavior with respect to `agent
status` output, as that behavior was not very intuitive and sometimes
incorrect.

In this commit, the container launcher includes its own, private notion
of a "tailer" which wraps existing logs-agent functionality: for
sockets, it wraps the socket tailer; for files, it wraps file _sources_.
Both of these wrappers are ugly and awful but allow reuse of the
existing functionality without modifying it.


### Motivation

This is part of the ongoing work to refactor the logs agent.

### Additional Notes

As with all `cca_in_ad` PRs, please begin by confirming that, if this config is false, nothing changes.   Then proceed to review the remainder.

This is a large PR that introduces a substantial chunk of new code, but with luck that new code is clear and well-structured so it's a joy to read.  If not, please provide constructive feedback :)

I intend to work on new tailers in subsequent PRs.  So, please limit reviews of the tailer "wrappers" in this PR to correctness -- the code won't live for very long.

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Release QA for this card (with `cca_in_ad: false`) is as simple as checking that a container can still be logged in any one scenario, since it changes nothing with that flag not set.

Non-release QA for this card includes verifying that, with `cca_in_ad: true`, behavior in the following scenarios is the same or better as with the flag not set:
 - docker daemon killed during logging
 - docker daemon stopped during logging, later restarted
 - docker_container_use_file but /var/lib/docker/containers not accessible -> fall back to socket
 - CCA + unlabeled container -> source/service based on image name
 - container with source+service set --> results in stream-logs
 - container with only service set --> results in stream-logs with default from image name
 - container with only source set --> results in stream-logs with default from image name
 - container with `[{}]` --> default source and service from image name
 - container with 'service:foo' standard tag --> service=foo, default source
 - pod with annotation including `{"type":"docker"}` -- should work for docker-based k8s, should be no worse than existing behavior for containerd-based k8s

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

djmitche committed:
cca_in_ad -> replace kubernetes/docker launchers with container launcher

The kubernetes and docker launchers have overlapping functionality, and
both have serious bugs and missing functionality.  The container
launcher implements the same behavior, with fewer bugs and more
functionality.  In particular, it

 - handles config from files (conf.d) for containers under kubernetes
 - handles changes to pod annotations
 - correctly logs from file or socket based on `logs_config.{k8s,docker}_container_use_file`
 - captures logs under containerd when `k8s_container_use_file` is false

The entire container launcher is hidden behind the 'docker' build tag,
which really means "containers" and not specifically docker.

Outside of the situations where they have obvious bugs, this launcher
attempts to preserve behavior of the kubernetes and docker launchers in
the details, such as defaults for the "Source" and "Service" properties.
It does not preserve those launchers' behavior with respect to `agent
status` output, as that behavior was not very intuitive and sometimes
incorrect.

In this commit, the container launcher includes its own, private notion
of a "tailer" which wraps existing logs-agent functionality: for
sockets, it wraps the socket tailer; for files, it wraps file _sources_.
Both of these wrappers are ugly and awful but allow reuse of the
existing functionality without modifying it.

djmitche commented:
(rebased to avoid lint errors on `main`)

gh123man commented on pkg/logs/internal/launchers/container/tailerfactory/file.go:
```diff
@@ -0,0 +1,260 @@
+		}
+
+	case containersorpods.LogPods:
+		return tf.makeK8sFileSource(source)
```
I think the K8s code needs to be split out from the file code. 
Currently all k8s code is behind the `kubelet` build tag. It looks like windows builds include the `docker` build tag but NOT the `kubelet` tag which is causing unsupported tests to run on windows (hence the failures)

gh123man commented on pkg/logs/internal/launchers/container/tailerfactory/file.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	// get the path for the discovered pod and container
+	path := findK8sLogPath(pod, container.Name)
```
UT's are filing on windows because all the paths are unix style. Interestingly, we seem to build the `kubelet` tag on windows even though I don't *think* we actually support k8s log collection on windows. 

I think we may need a new build tag to seperate out k8s logging behavior from the `kubelet` since it seems the support matricx for windows does not include logging. 
It seems like in the past we were just getting lucky that the tests worked on windows and the code was never actually called on a real agent - because I can't imagine it would work (unless I'm missing something)

gh123man replied:
Actually - I confirmed that logs on windows is fully supported (or it at least works today)

gh123man replied:
Ive pushed a fix for the windows UTs. 
If CI passes will merge

gh123man committed:
Merge branch 'main' into dustin.mitchell/ac-1644-container-launcher

gh123man committed:
fix windows UTs

sgnn7 reviewed (approved):
LGTM for commit `a6239e11e40322fdffc8ae92cede2eae942c9c4f`

gh123man committed:
linter

gh123man committed:
Merge branch 'main' into dustin.mitchell/ac-1644-container-launcher

djmitche committed:
Use backslashes when testing filenames on Windows

This reverts commit 6149317e1c3d91cb261f56d51c60b8996258aa12.

djmitche commented on pkg/logs/internal/launchers/container/tailerfactory/factory.go:
```diff
@@ -0,0 +1,103 @@
+		if err == nil {
+			return t, nil
+		}
+		log.Warnf("Could not make socket tailer for source %s (falling back to file): %v", source.Name, err)
```
The docker launcher doesn't actually retry connecting to the docker socket, so the alternative is not logging that container at all (until agent restart).  The kubernetes launcher [used to](https://github.com/DataDog/datadog-agent/commit/139413ec4c5e5c55f1d76e45d8847be409a8793c#diff-6db01b65594d5d28f0fa2d81c9f18cff2269109fbd6cd7f3f3fd36ce3a50117eL40) retry when kubeutil wasn't available, but that was recently removed.  I don't think we support transient connection problems to the docker socket.

djmitche replied:
Oh, and we _do_ fall back to files, at least in a k8s context:
 * agent starts, is unable to access docker but able to access kubelet
 * LogWhat = LogPods (regardless of k8s_container_use_file)
 * kubernetes launcher starts (so docker_container_use_file is not relevant)
 * kubernetes launcher only supports file logging

djmitche commented:
I don't think the build jobs are failing due to this PR -- I'll see if a merge helps.

djmitche committed:
Merge remote-tracking branch 'upstream/main' into dustin.mitchell/ac-1644-container-launcher

djmitche commented:
The yellow jobs are due to issues syncing from GitLab - the pipeline is green.

djmitche committed:
ensure SourceType is correct for child file sources

djmitche committed:
container-launcher: get image info from WLM (#12637)

Using dockerutil to get container information assumes that it is a
docker container, which is not always the case. WLM can provide the
information directly and more simply.

djmitche commented:
(QA steps here already cover those of #12637)
